### PR TITLE
[INTEL MKL] Bug-fix to in-place computation with tensor forwarding.

### PR DESCRIPTION
**Pull request by mdfaijul**
This PR fixes a bug in a fused operator (`conv2d` + `bias_add` + `add` + `relu`), wherein one of the inputs to `add` is used as in-place buffer for output result. Previous code was forcefully forwarding the input to the output, although the input could have more than one fan-outs.

### CODE REVIEW

File: tensorflow/python/kernel_tests/conv_ops_test.py [hunk 1]
```diff
@@ -3266,6 +3267,164 @@ def Test(self):
...
+    conv = self._CreateConv2D(conv1, filter_in - ofs)
+    bias_add = nn_ops.bias_add(conv, bias_in)
+    add = math_ops.add_n([bias_add, conv2])
+
+    self.assertAllEqual(
+        np.rint(expected_output),
+        self.evaluate(add).reshape(-1))
+
+  @test_util.deprecated_graph_mode_only
+  def testAddWithRefCountTwoAndRunAddLast(self):
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/conv_ops_test.py:3314]: The comment here could be:
```
# Tests tensor forwarding of a fused Conv2D+BiasAdd+Add op when the input to Add
# has a total refcount of 2, and Add is its last consumer.
```

File: tensorflow/python/kernel_tests/conv_ops_test.py [hunk 2]
```diff
@@ -3266,6 +3267,164 @@ def Test(self):
...
+    conv = self._CreateConv2D(conv2, filter_in - ofs)
+    bias_add = nn_ops.bias_add(conv, bias_in)
+    add = math_ops.add_n([bias_add, conv1])
+
+    self.assertAllEqual(
+        np.rint(expected_output),
+        self.evaluate(add).reshape(-1))
+
+  @test_util.deprecated_graph_mode_only
+  def testAddWithRefCountTwoAndRunAddFirst(self):
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/conv_ops_test.py:3343]: The comment here could be:
```
# Tests tensor forwarding of a fused Conv2D+BiasAdd+Add op when the input to Add
# has refcount 2 and Add (in the fused Conv2D op) is its first consumer.
```
Or we could use other op than Add in `output = math_ops.add_n([relu, conv2])` and remove the `(in the fused Conv2D op)` part in the comment.

File: tensorflow/python/kernel_tests/conv_ops_test.py [hunk 3]
```diff
@@ -3266,6 +3267,164 @@ def Test(self):
...
+
+    relu = nn_ops.relu(add)
+    output = math_ops.add_n([relu, conv2])
+
+    self.assertAllEqual(
+        np.rint(expected_output),
+        self.evaluate(output).reshape(-1))
+
+  @test_util.deprecated_graph_mode_only
+  def testAddWithRefCountTwoAndNoDependence(self):
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/conv_ops_test.py:3373]: The comment here could be:
```
# Tests tensor forwarding of a fused Conv2D+BiasAdd+Add op when the input to Add
# has refcount 2, and there is no dependency between its two consumers.
```

File: tensorflow/python/kernel_tests/conv_ops_test.py [hunk 4]
```diff
@@ -3266,6 +3267,164 @@ def Test(self):
...
+    relu2 = nn_ops.relu(conv2)
+    output = math_ops.add_n([relu1, relu2])
+
+    self.assertAllEqual(
+        np.rint(expected_output),
+        self.evaluate(output).reshape(-1))
+
+
+  @test_util.deprecated_graph_mode_only
+  def testAddWithSameSrcAndAddTensorBuffer(self):
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/conv_ops_test.py:3405]: The comment here could be:
```
# Tests tensor forwarding of a fused Conv2D+BiasAdd+Add op when the input to Add
# is the same as the input to the fused Conv2D op and needs a tensor buffer.
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and sorry for the delay!

File: tensorflow/core/util/mkl_util.h [hunk 5]
```diff
@@ -893,20 +893,36 @@ inline void SetDummyMklDnnShapeOutput(OpKernelContext* context,
...
-    context->set_output(idx_data_out, context->input(idx_data_in));
+    is_forwarded = context->forward_input_to_output_with_shape(
+        idx_data_in, idx_data_out, output_shape, output);
+  }
+  if (is_forwarded || always_forward) {
+    AllocateOutputSetMklShape(context, idx_out, mkl_shape);
+    return true;
+  } else {
+    return false;
   }
```
[the remark this answers is not in the document]
Comment (mdfaijul) [reply tensorflow/core/util/mkl_util.h]: Done.

File: tensorflow/core/util/mkl_util.h [hunk 6]
```diff
@@ -893,20 +893,36 @@ inline void SetDummyMklDnnShapeOutput(OpKernelContext* context,
   AllocateOutputSetMklShape(context, idx_data_out, mkl_shape_output);
 }
 
-inline void ForwardMklTensorInToOutWithMklShape(OpKernelContext* context,
+// If the input tensor has ref count as 1, it is forwarded to the desired
+// output port and the function reutrns true. In that case, it also allocates
```
[the remark this answers is not in the document]
Comment (mdfaijul) [reply tensorflow/core/util/mkl_util.h]: Done

File: tensorflow/python/kernel_tests/conv_ops_test.py [hunk 7]
```diff
@@ -3266,6 +3267,164 @@ def Test(self):
   return Test
 
 
+class FusedConv2DTest(test.TestCase):
+
+  def _CreateNumpyTensor(self, shape):
+    total_size = 1
+    for s in shape:
+      total_size *= s
+    return np.arange(1, total_size + 1, dtype=np.float32).reshape(shape)
```
[the remark this answers is not in the document]
Comment (mdfaijul) [reply tensorflow/python/kernel_tests/conv_ops_test.py:3274]: Done

File: tensorflow/python/kernel_tests/conv_ops_test.py [hunk 8]
```diff
@@ -3266,6 +3267,164 @@ def Test(self):
...
+
+  def _CreateConv2D(self, input_values, filters,
+                    strides=[1, 1], padding="SAME"):
+    return nn_ops.convolution(
+              input_values,
+              filters,
+              strides=strides,
+              padding=padding)
+
+  @test_util.deprecated_graph_mode_only
```
[the remark this answers is not in the document]
Comment (mdfaijul) [reply tensorflow/python/kernel_tests/conv_ops_test.py]: Done

File: tensorflow/python/kernel_tests/conv_ops_test.py [hunk 9]
```diff
@@ -3266,6 +3267,164 @@ def Test(self):
...
+        4783, 6366, 779, 1134]
+    tensor_in_sizes = [1, 3, 3, 2]
+    filter_in_sizes = [2, 2, 2, 2]
+    bias_in_sizes = [2]
+
+    x = self._CreateNumpyTensor(tensor_in_sizes)
+    filter_in = self._CreateNumpyTensor(filter_in_sizes)
+    bias_in = self._CreateNumpyTensor(bias_in_sizes)
+    # To get different weights for filter
+    ofs = 1
```
[the remark this answers is not in the document]
Comment (mdfaijul) [reply tensorflow/python/kernel_tests/conv_ops_test.py]: Done

File: tensorflow/python/kernel_tests/conv_ops_test.py [hunk 10]
```diff
@@ -3266,6 +3267,164 @@ def Test(self):
...
+                    strides=[1, 1], padding="SAME"):
+    return nn_ops.convolution(
+              input_values,
+              filters,
+              strides=strides,
+              padding=padding)
+
+  @test_util.deprecated_graph_mode_only
+  def testAddWithRefCountOne(self):
+    expected_output = [
```
[the remark this answers is not in the document]
Comment (mdfaijul) [reply tensorflow/python/kernel_tests/conv_ops_test.py:3288]: Done

File: tensorflow/python/kernel_tests/conv_ops_test.py (hunk 1 above)
Comment (mdfaijul) [reply tensorflow/python/kernel_tests/conv_ops_test.py:3316]: Done

File: tensorflow/python/kernel_tests/conv_ops_test.py (hunk 2 above)
Comment (mdfaijul) [reply tensorflow/python/kernel_tests/conv_ops_test.py:3347]: Done

File: tensorflow/python/kernel_tests/conv_ops_test.py (hunk 3 above)
Comment (mdfaijul) [reply tensorflow/python/kernel_tests/conv_ops_test.py:3379]: Done

File: tensorflow/python/kernel_tests/conv_ops_test.py (hunk 4 above)
Comment (mdfaijul) [reply tensorflow/python/kernel_tests/conv_ops_test.py:3413]: Done

**Comment by mdfaijul** [comment]
@penpornk Thanks a lot for the valuable comments. I have addressed them. Please check.

**Review by penpornk — APPROVED** [review]
Thank you for the quick changes!

**Review by rthadur — APPROVED** [review]

**Pull request merged**